### ISSUE TITLE
More Real Unit Tests

### ISSUE THREAD

**Issue by kevinkreiser** [valhalla/valhalla#1170]
At this point we have been carrying around a small extract of Utrecht, NL for a while now. This allows us to fully test most of the actions in the code. While unit tests aren't generally the place for doing full scale feature level validation, having the safety net there has been really valuable. I would like us to add a small test suite that tests all of the actions and the options into those actions at a high level. Basically:

```
/route
/optimized_route
/trace_*
/locate
/sources_to_targets
/isochrones
```

and test different permutations of options to see that they have the desired affect. some of these are harder to validate than others but i think widening the safety net is a good goal for us to have.

i would propose that we find very small routes (for the sake of speed of testing and simplicity of debugging). and then change options on those to see they have the desired effects. this will go a long way to avoiding regressions without having to run a full integration test. basically catch the low hanging fruit and make refactorings like the kind we've just had (are still having) feel less risky.

**Comment by dnesbitt61** [comment]
The only issue I have with testing on this static data set is how to visualize the data and tagging. Using a snapshot of OSM makes it so the test behavior remains consistent but makes it hard to produce test cases as time goes by and OSM data changes.

**Comment by kevinkreiser** [comment]
i guess it would depend on what you mean by visualize exactly...

if you mean render raster tiles with the data, we can do that but we dont have it today. while doing this is not new territory for us, it would be a moderate effort of cobbling together some scripts with osm2pgsql and mapnik. something like https://wiki.openstreetmap.org/wiki/Creating_your_own_tiles would be what we are looking for. the thing about static data though, is we only need to do it once and just keep the tiles somewhere or document how we did it

if you mean looking at the tagging we kind of get that for free by doing the rendering. since to do the rendering you probably loaded the data into a database and the thing about databases is that they make it easy to get information. if you didnt go to the trouble of loading it into a database we can still, easily convert it to plaintext xml. and just grep for the ways or nodes or relations in question by id.

i wouldn't put these as pre-requisites for adding the basic testing though and here is why: a lot of the functions that i would like to get regression testing for dont rely on explicit features of the data itself. like im going to be testing that the speed for a given edge from `trace_attributes` is the correct value from the data. i am more worried about stuff like i asked for the speeds and the response contains them, i dont care what the value is because as you said its data dependent.

basically i was just looking for something a little more pie in the sky to keep us covered from regressions at the API level not at the data level. i dont even care what path a given algorithm is taking. i just care that the proper outputs are there not what values they hold. thta is why i said just small routes (like 3 edges or something). i dont want to check the correctness of the algorithm, for that we should have other unit tests. i want to check that the service API isnt having breaking changes. am i making sense?
